### PR TITLE
[JUJU-1825] Secret k8s backend

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR implements the secret k8s backend based on k8s RBAC sources.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```console
$ juju add-model t1 --config secret-store=kubernetes

$ juju deploy hello-kubecon
Located charm "hello-kubecon" in charm-hub, revision 14
Deploying "hello-kubecon" from charm-hub charm "hello-kubecon", revision 14 in channel stable on ubuntu/20.04

$ juju deploy nginx-ingress-integrator
Located charm "nginx-ingress-integrator" in charm-hub, revision 29
Deploying "nginx-ingress-integrator" from charm-hub charm "nginx-ingress-integrator", revision 29 in channel stable on ubuntu/20.04

$ juju relate hello-kubecon nginx-ingress-integrator

$ juju trust nginx-ingress-integrator --scope=cluster

$ juju exec --unit hello-kubecon/0  "secret-add --owner=unit foo=bar"
secret:ccv33c6ffbahq9l9ie0g

$ juju exec --unit hello-kubecon/0 "secret-get secret:ccv33c6ffbahq9l9ie0g"
foo: bar

$ mkubectl -nt1 describe role.rbac.authorization.k8s.io/unit-hello-kubecon-0
Name:         unit-hello-kubecon-0
Labels:       app.kubernetes.io/managed-by=juju
              app.kubernetes.io/name=hello-kubecon
Annotations:  <none>
PolicyRule:
  Resources     Non-Resource URLs  Resource Names            Verbs
  ---------     -----------------  --------------            -----
  secrets.*     []                 [ccv33c6ffbahq9l9ie0g-1]  [*]
  secrets.*     []                 []                        [create patch]
  namespaces.*  []                 [t1]                      [get list]

$ juju exec --unit hello-kubecon/0 "secret-grant -r 0 secret:ccv33c6ffbahq9l9ie0g"

$ juju exec --unit nginx-ingress-integrator/0 "secret-get secret:ccv33c6ffbahq9l9ie0g"
foo: bar

$ mkubectl -nt1 describe role.rbac.authorization.k8s.io/unit-nginx-ingress-integrator-0
Name:         unit-nginx-ingress-integrator-0
Labels:       app.kubernetes.io/managed-by=juju
              app.kubernetes.io/name=nginx-ingress-integrator
Annotations:  <none>
PolicyRule:
  Resources     Non-Resource URLs  Resource Names            Verbs
  ---------     -----------------  --------------            -----
  secrets.*     []                 []                        [create patch]
  namespaces.*  []                 [t1]                      [get list]
  secrets.*     []                 [ccv33c6ffbahq9l9ie0g-1]  [get]

# add another secret.
$ juju exec --unit hello-kubecon/0  "secret-add --owner=unit foo=2bar"
secret:ccv3eauffbahq9l9ie1g

$ juju exec --unit hello-kubecon/0 "secret-get secret:ccv3eauffbahq9l9ie1g"
foo: 2bar

# not grant yet.
$ juju exec --unit nginx-ingress-integrator/0 "secret-get secret:ccv3eauffbahq9l9ie1g"
ERROR permission denied
ERROR the following task failed:
 - id "26" with return code 1

use 'juju show-task' to inspect the failure


```

## Documentation changes

No

## Bug reference

No

### CODE REVIEW

File: caas/kubernetes/provider/application/application.go [hunk 1]
```diff
@@ -1416,7 +1416,7 @@ func (a *app) ApplicationPodSpec(config caas.ApplicationConfig) (*corev1.PodSpec
 		AutomountServiceAccountToken:  &automountToken,
 		ServiceAccountName:            a.serviceAccountName(),
 		ImagePullSecrets:              imagePullSecrets,
-		TerminationGracePeriodSeconds: pointer.Int64Ptr(300),
+		TerminationGracePeriodSeconds: pointer.Int64Ptr(10),
```
Comment (ycliuhw) [inline caas/kubernetes/provider/application/application.go:1419]: Changed for quicker testing, revert later.

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 2]
```diff
@@ -219,7 +220,8 @@ func (s *SecretsManagerAPI) RemoveSecrets(args params.DeleteSecretArgs) (params.
 	if err != nil {
 		return params.ErrorResults{}, errors.Trace(err)
 	}
-	if err := provider.CleanupSecrets(model, removedURIs); err != nil {
+	// TODO: include unitTag in params.DeleteSecretArgs for operator uniters?!!!
```
Comment (ycliuhw) [inline apiserver/facades/agent/secretsmanager/secrets.go:223]: Currently, 3.0 is broken the operator style applications due to https://bugs.launchpad.net/juju/+bug/1991213
This PR won't cover the legacy operator-style applications.
We need to revisit once lp:1991213 is fixed.

File: caas/kubernetes/provider/application/trust.go [hunk 3]
```diff
@@ -12,23 +12,33 @@ import (
...
 
-var defaultApplicationNamespaceRules = []rbacv1.PolicyRule{
-	{
-		APIGroups: []string{""},
-		Resources: []string{"pods", "services"},
-		Verbs: []string{
-			"get",
-			"list",
-			"patch",
+func getDefaultApplicationNamespaceRules(namespace string) []rbacv1.PolicyRule {
```
Comment (ycliuhw) [inline caas/kubernetes/provider/application/trust.go:15]: This change will disappear once rebase is done after https://github.com/juju/juju/pull/14699 merged.

File: caas/kubernetes/provider/modeloperator.go [hunk 4]
```diff
@@ -693,6 +693,25 @@ func ensureExecRBACResources(objMeta meta.ObjectMeta, clock jujuclock.Clock, bro
 	role := &rbac.Role{
 		ObjectMeta: objMeta,
 		Rules: []rbac.PolicyRule{
+			{
+				APIGroups: []string{""},
```
Comment (ycliuhw) [inline caas/kubernetes/provider/modeloperator.go:697]: This change will disappear once rebase is done after https://github.com/juju/juju/pull/14699 merged

File: caas/kubernetes/provider/application/application.go (hunk 1 above)
Comment (wallyworld) [reply caas/kubernetes/provider/application/application.go:1419]: Don't forget to revert before landing

File: secrets/provider/provider_test.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+var _ = gc.Suite(&providerSuite{})
+
+func (*providerSuite) TestNameMetaSlice(c *gc.C) {
+	nameMeta := provider.NameMetaSlice{}
+	nameMeta.Add(&secrets.URI{ID: "a"}, 1)
+	nameMeta.Add(&secrets.URI{ID: "b"}, 1, 2)
+	nameMeta.Add(&secrets.URI{ID: "c"}, 1, 2, 3)
+	nameMeta.Add(&secrets.URI{ID: "d"}, 1, 2, 3, 4)
```
Comment (wallyworld) [inline secrets/provider/provider_test.go:25]: add revisions to "d" in 2 steps to check append

File: secrets/provider/vault/secrets.go [hunk 6]
```diff
@@ -13,27 +13,33 @@ import (
...
 )
 
+var logger = loggo.GetLogger("juju.secrets.vault")
+
 const (
 	// Store is the name of the Kubernetes secrets store.
 	Store = "vault"
 )
 
 // NewProvider returns a Kubernetes secrets provider.
```
Comment (wallyworld) [inline secrets/provider/vault/secrets.go:31]: typo

File: secrets/provider/provider.go [hunk 7]
```diff
@@ -18,25 +23,51 @@ type Model interface {
 	UUID() string
 }
 
+// NameMetaSlice stores information for generating resource name for a list of secrets.
+type NameMetaSlice map[string]set.Ints
```
Comment (wallyworld) [inline secrets/provider/provider.go:27]: it's more correct to call it "SecretRevisions"?

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 8]
```diff
@@ -203,12 +204,12 @@ func (s *SecretsManagerAPI) RemoveSecrets(args params.DeleteSecretArgs) (params.
 	result := params.ErrorResults{
 		Results: make([]params.ErrorResult, len(args.Args)),
 	}
-	var removedURIs []*coresecrets.URI
+	removedURIs := secretsprovider.NameMetaSlice{}
```
Comment (wallyworld) [inline apiserver/facades/agent/secretsmanager/secrets.go:207]: it's really "removedRevisions" now right?

File: core/secrets/secret.go [hunk 9]
```diff
@@ -91,6 +91,10 @@ func NewURI() *URI {
 	}
 }
 
+func (u URI) Name(revision int) string {
```
Comment (wallyworld) [inline core/secrets/secret.go:94]: comment

File: secrets/provider/kubernetes/secrets.go [hunk 10]
```diff
@@ -42,45 +54,119 @@ func (p k8sProvider) CleanupModel(m provider.Model) error {
...
+	}
+	cred, err := k8scloud.UpdateCredentialWithToken(*cloudSpec.Credential, token)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cloudSpec.Credential = &cred
+
+	if cloudSpec.IsControllerCloud {
+		// the cloudspec used for controller having fake endpoint(address and port)
+		// because we ignores these endpoint and loads in-cluster credential instead.
```
Comment (wallyworld) [inline secrets/provider/kubernetes/secrets.go:145]: typo

File: secrets/provider/kubernetes/secrets.go [hunk 11]
```diff
@@ -156,5 +245,5 @@ func (k k8sStore) DeleteContent(ctx context.Context, providerId string) error {
 
 // SaveContent implements SecretsStore.
 func (k k8sStore) SaveContent(ctx context.Context, uri *secrets.URI, revision int, value secrets.SecretValue) (string, error) {
-	return k.broker.SaveJujuSecret(ctx, uri, revision, value)
+	return k.broker.SaveJujuSecret(ctx, uri.Name(revision), value)
```
Comment (wallyworld) [inline secrets/provider/kubernetes/secrets.go:248]: uri.Name() is only used once I think. And elsewhere we still use fmt.Printf("%s-%s"), eg in metaslice Names()
So perhaps the extra method doesn't really pay off and we can do without it.

**Review by wallyworld — APPROVED** [review]
Just a few small tweaks. Thanks very much for adding the missing tests.
It would be good to get those bugs that were mentioned fixed asap so we can get the interfaces all stable before release.

File: secrets/provider/kubernetes/secrets.go (hunk 11 above)
Comment (ycliuhw) [reply secrets/provider/kubernetes/secrets.go:248]: Both `k8sStore.SaveContent` and `vaultStore.SaveContent` uses this. And future providers should use this as well.

Comment (ycliuhw) [reply secrets/provider/kubernetes/secrets.go:248]: Now we always use `uri.Name` to get the secret resource name.

**Pull request merged**